lib/grp.gi

```diff
@@ -4291,86 +4294,121 @@ end );
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupType;
-    else
-      if not IsBound(fam!.defaultGroupType) then
-        fam!.defaultGroupType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses);
+function( gens )
+local G,fam,typ,id,triv;
+
+  fam:=FamilyObj(gens);
+
+  triv:=fail; # might not find out
+
+  typ:=IsGroup and IsAttributeStoringRep 
+    and HasIsEmpty and HasGeneratorsOfMagmaWithInverses;
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    if Length(gens)<=1 then
+      typ:=typ and HasIsCyclic and IsCyclic 
+          and HasIsCommutative and IsCommutative;
+    fi;
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      id:=One(gens[1]);
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=false;
+      else
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=true;
+        # basic consequence
+        typ:=typ and HasIsPerfectGroup and IsPerfectGroup;
       fi;
-      typ:=fam!.defaultGroupType;
     fi;
 
-    G:=rec();
-    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
+  fi;
+  typ:=NewType(fam,typ);
 
-    return G;
-    end );
+  G:=rec();
+  if triv=fail then
+    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens),
+                            IsEmpty,false);
+  else
+    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens),
+                            IsEmpty,false,IsTrivial,triv,
+                            IsNonTrivial,not triv);
+  fi;
+
+  return G;
+end );
 
 InstallMethod( GroupWithGenerators,
     "generic method for collection and identity element",
     IsCollsElms, [ IsCollection, IsMultiplicativeElementWithInverse ],
-    function( gens, id )
-    local G,fam,typ;
-
-    fam:=FamilyObj(gens);
-    if IsFinite(gens) then
-      if not IsBound(fam!.defaultFinitelyGeneratedGroupWithOneType) then
-        fam!.defaultFinitelyGeneratedGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses
-                      and IsFinitelyGeneratedGroup and HasOne);
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupWithOneType;
-    else
-      if not IsBound(fam!.defaultGroupWithOneType) then
-        fam!.defaultGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses and HasOne);
+function( gens, id )
+local G,fam,typ,triv;
+
+  fam:=FamilyObj(gens);
+  triv:=fail; # might not find out
+
+  typ:=IsGroup and IsAttributeStoringRep 
+        and HasIsEmpty and HasGeneratorsOfMagmaWithInverses and HasOne;
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    if Length(gens)<=1 then
+      typ:=typ and HasIsCyclic and IsCyclic 
+          and HasIsCommutative and IsCommutative;
```

Comment:
Note that this is equivalent to `typ := typ and IsCyclic`, as `IsCyclic` implies `IsCommutative`, and each implies its tester, and handling these implications comes at zero cost here.

Reply:
Ah -- I wasn't sure. Will fix.

Reply:
Every property implies it tester; for the rest, you can check this e.g. as follows :
```
gap> ShowImpliedFilters(IsCyclic);


May imply with:
+IsMagmaWithInverses
+IsAssociative
   IsCommutative
   IsNilpotentGroup
   IsSupersolvableGroup
   IsSolvableGroup
   IsNilpotentByFinite

+IsOrdinaryTable
   IsCommutative
   IsNilpotentCharacterTable
   IsSolvableCharacterTable
   IsSupersolvableCharacterTable
```

Or, more technically:
```
gap> IsEquivFilters:={f1,f2}->IS_EQUAL_FLAGS(WITH_IMPS_FLAGS(FLAGS_FILTER(f1)), WITH_IMPS_FLAGS(FLAGS_FILTER(f2)));
function( f1, f2 ) ... end
gap> IsEquivFilters(IsGroup and IsCyclic, HasIsCyclic and IsCyclic and HasIsCommutative and IsCommutative and IsGroup);
true
```